### ISSUE TITLE
pfSense `/var/lib/tailscale` is volatile, state should be stored somewhere else

### ISSUE THREAD

**Issue by stratus-ss** [tailscale/tailscale#6861]
### What is the issue?

1) I created a tag as per the documentation
2) create an auth key toggling the tags and setting the `tag:server`
3) run the following
```
/usr/local/bin/tailscale up --advertise-exit-node  --advertise-routes "192.168.x.0/24,192.168.x.0/24" --auth-key xxxx
```

At this point I expect to be logged in. Instead I receive the following errors

```
backend error: invalid key: API key k3uvgP1CNTRL not valid
```
If I check the tailscale console, I see the key has been moved to revoked

### Steps to reproduce

The tailscale UI does not function on fresh install of pfSense 2.6
I need to drop down to the CLI in order to start tailscale

### Are there any recent changes that introduced the issue?

as far as I know this has always happened

### OS

Other

### OS version

pfSense 2.6

### Tailscale version

tailscale --version 1.26.2   go version: go1.18.1

### Bug report

BUG-fa2228616e0418c3dc43442ae2c2be8fc76ff13c8855e5fb2b1a25e300215537-20221231043008Z-0f70cd0cc60c563b

**Comment by DentonGentry** [comment]
It looks like the UI had already tried to bring the tailnet up, we see it receiving netmap updates:
```
2022-12-31 04:31:06.017329204 +0000 UTC: magicsock: got updated network map; 4 peers
2022-12-31 04:31:06.017747462 +0000 UTC: wgcfg: skipped unselected default routes from: "XXX" ([GGCdZ])
```

You may need to use `/usr/local/bin/tailscale up --advertise-exit-node  --advertise-routes "192.168.x.0/24,192.168.x.0/24" --auth-key xxxx --force-reauth`, adding the force-reauth flag to make it give up its current node identity and authenticate again.

**Comment by stratus-ss** [comment]
thanks for the hint.

When I tried to use the UI, I always see this:

![image](https://user-images.githubusercontent.com/8162705/210138772-63070e7c-b69f-4093-8570-fef56220633a.png)

What I am trying to do is get tailscale to start on boot. Since the UI always shows the above error I was trying to create a startup script.

With this in mind is the `--force-reauth` appropriate for a startup script or should it only be used once?

**Comment by DentonGentry** [comment]
--force-reauth would only be used once, tailscale will store its new identity in the state file. --force-reauth does not need to be used on every boot.

Alternately: if the goal is just to get the pfSense device to be tagged, in https://login.tailscale.com/admin/machines in the three dot menu to the right of the entry for the machine is an "Edit ACL Tags..." item which can apply tags. You'll likely also want to select the "Disable key expiry" menu selection for infrastructure like pfSense.

**Comment by stratus-ss** [comment]
> --force-reauth would only be used once, tailscale will store its new identity in the state file. --force-reauth does not need to be used on every boot.
> 
> Alternately: if the goal is just to get the pfSense device to be tagged, in https://login.tailscale.com/admin/machines in the three dot menu to the right of the entry for the machine is an "Edit ACL Tags..." item which can apply tags. You'll likely also want to select the "Disable key expiry" menu selection for infrastructure like pfSense.

I really appreciate the responses, the problem I am trying to solve is that tailscale does not start on a reboot. I had initially started with the method you mentioned with the manual auth and disabled the key but the UI does not seem to actually make tailscale connections. It starts the daemon but thats it. 

Previously I would just log into the shell and do `tailscale up`. However, if I am trying to use this as a "vpn" situation the intent is to be remote and I can't bring up the tailnet remotely in this case


Also note that when I used the `--force-reauth` on the cli, it connects but the key is revoked immediately

![image](https://user-images.githubusercontent.com/8162705/210138883-b86d65af-ef2a-4679-9129-0efe04e4845c.png)

**Comment by DentonGentry** [comment]
A single-use key will be marked as Revoked when it is used. So far as I can see, it did authorize the node.

**Comment by stratus-ss** [comment]
The problem i am experiencing is that regardless of the method I use to set up tailscale on PFSense it is constantly asking for authorization. If i follow the same pattern on an Ubuntu machine I have no problems. For example I installed an Ubuntu VM, I ran through the same CLI command to start up, instead of using the auth-key I let it generate the URL and then went to the URL to authorize the device

The PFSense box deauthorizes/requires reauth on reboot where the ubuntu box doesn't

**Comment by stratus-ss** [comment]
The problem i am experiencing is that regardless of the method I use to set up tailscale on PFSense it is constantly asking for authorization. If i follow the same pattern on an Ubuntu machine I have no problems. For example I installed a CentOS Stream VM, I ran through the same CLI command to start up, instead of using the auth-key I let it generate the URL and then went to the URL to authorize the device

The PFSense box deauthorizes/requires reauth on reboot where the CentOS Stream box doesn't.

**Comment by DentonGentry** [comment]
It is behaving like the `tailscaled.state` file has been stored somewhere volatile and is being lost at each reboot.

**Comment by Doable0113** [comment]
I'm having the same issue with pfsense, that I have to get a new pre-authentication key after every reboot.  I have use RAM disks enabled and that may be causing the problem of the tailscaled.state file being lost.

**Comment by stratus-ss** [comment]
I guess the question is, is it possible to reconfigure tailscale on PFsense to move the state and configuration files? like @Doable0113 mentions its likely due to /var being a ram disk. I don't want to have to disable ram disk on /var just to get tailscale working

**Comment by DentonGentry** [comment]
We can make the default state storage differ by platform. Where would pfsense usually store persistent state? `/etc`?

**Comment by stratus-ss** [comment]
It is common on PFSense to have configurations and state stored in `/usr/local/etc`.

For example, LDAP, Unbound, Telegraph and so on, store their configs and related files there

**Comment by Jellman01** [comment]
For me, a similar issue is caused by using the ramdisk option in PFsence. The tunnel is lost on reboot and will not reconnect unless you generate a new key.

Disabling the RAM Disk fixes the issue. It would be great if this could be fixed.

**Comment by achtnullzwei** [comment]
Remembered this issue here and just wanted to highlight that in the recent Tailscale package on pfsense this should be fixed. They also made the location of the state file configurable in the GUI. 

Hope its fine the link here to their redmine page for reference: https://redmine.pfsense.org/issues/13632

**Comment by stratus-ss** [comment]
thanks for the update! I will definitely look into this now and report back

**Comment by alirz1** [comment]
Same here. Waiting for this to be fixed in pfsense CE 2.6

**Comment by DentonGentry** [comment]
This is believed to be fixed in pfSense 23.05-BETA and later, according to https://redmine.pfsense.org/issues/13632
